File: frontend/lib/api/activity-logs.ts
import { apiFetch } from "./client";

export interface ActivityLog {
  id: string;
  event_type: string;
  entity_type: string;
  entity_id?: string | null;
  patient_id?: string | null;
  appointment_id?: string | null;
  actor_type: string;
  description?: string | null;
  metadata?: Record<string, unknown> | null;
  created_at: string;
}

/** GET /activity-logs?limit= */
export async function getActivityLogs(limit: number = 10): Promise<ActivityLog[]> {
  return apiFetch<ActivityLog[]>(`/activity-logs?limit=${limit}`);
}

File: src/api_service/routers/activity_logs.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.database.base_engine import get_db
from src.database.models import ActivityLog


router = APIRouter(
    prefix="/activity-logs",
    tags=["activity-logs"],
)


@router.get("")
async def get_activity_logs(
    limit: int = Query(default=20, ge=1, le=100),
    patient_id: UUID | None = None,
    appointment_id: UUID | None = None,
    session: AsyncSession = Depends(get_db),
):
    stmt = (
        select(ActivityLog)
        .order_by(ActivityLog.created_at.desc())
        .limit(limit)
    )

    if patient_id:
        stmt = stmt.where(ActivityLog.patient_id == patient_id)

    if appointment_id:
        stmt = stmt.where(
            ActivityLog.appointment_id == appointment_id
        )

    result = await session.execute(stmt)

    return result.scalars().all()